### PR TITLE
Support for non u64 values inside window frames

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Support for non-u64 types for Window Bound

# Which issue does this PR close?

This PR closes the #3571 . 

# Rationale for this change

With this change we are adding support for queries like below.

```sql
SELECT 
	COUNT(*) OVER(ORDER BY ts RANGE BETWEEN '1 DAY' PRECEDING AND '1 DAY' FOLLOWING) as cnt 
FROM t
```

# What changes are included in this PR?

- Closing this issue requires support for parsing expressions inside window frame queries. Hence we have updated sqlparser version.
- `assert_contains` and `assert_not_contains` macros are used in tests at different places. These macros were duplicated where they were used, we have moved them under [https://github.com/synnada-ai/arrow-datafusion/blob/feature/window_bound_scalar/datafusion/common/src/test_util.rs](https://github.com/synnada-ai/arrow-datafusion/blob/feature/window_bound_scalar/datafusion/common/src/test_util.rs) to prevent code duplication.
- Util codes for datetime arithmetic are moved under [https://github.com/synnada-ai/arrow-datafusion/blob/feature/window_bound_scalar/datafusion/common/src/scalar.rs](https://github.com/synnada-ai/arrow-datafusion/blob/feature/window_bound_scalar/datafusion/common/src/scalar.rs). They are reached now with `.add` and `.sub` api of `ScalarValue` enum.
- We removed type coercions between `ScalarValue` types. Arithmetic operations are done only between same types (for `datetime` types arithmetic operations are done only with `INTERVAL` types.) 
- Previously we were rejecting queries in the form

```sql
SELECT
      COUNT(c1) OVER (ORDER BY c2 RANGE BETWEEN 1 PRECEDING AND 2 PRECEDING)
      FROM aggregate_test_100
```

during window frame creation. Since we are supporting more reach variants such as 

```sql
SELECT 
	COUNT(*) OVER(ORDER BY ts RANGE BETWEEN '1 DAY' PRECEDING AND '1 MONTH' PRECEDING) as cnt 
FROM t
```

it is not easy to do rejection during creation. We now do calculations for whether window frame is valid after physical schema information is available. Where we can do validation calculations with appropriate types. This doesn’t change anything for end users.

- sqlparser had also changes in `Expr` type. We have incorporated those changes. However, possibly they are written with some new capabilities in mind. We just did the bare minimum to support the current test suit.

# Are there any user-facing changes?

N.A

### CODE REVIEW

**Comment by mustafasrepo** [comment]
test

**Comment by alamb** [comment]
I will attempt to make a sql parser release in the next few days

**Comment by alamb** [comment]
I released sqlparser 🎉

File: datafusion/optimizer/src/type_coercion.rs [hunk 1]
```diff
@@ -410,11 +413,121 @@ impl ExprRewriter for TypeCoercionRewriter {
...
+                };
+                Ok(expr)
+            }
             expr => Ok(expr),
         }
     }
 }
 
+/// Casts the ScalarValue `value` to coerced type.
+// When coerced type is `Interval` we use `parse_interval` since `try_from_string` not
```
Comment (alamb) [inline datafusion/optimizer/src/type_coercion.rs:440]: 👍

File: datafusion/sql/src/utils.rs [hunk 2]
```diff
@@ -480,6 +480,16 @@ pub fn window_expr_common_partition_keys(window_exprs: &[Expr]) -> Result<&[Expr
         .iter()
         .map(|expr| match expr {
             Expr::WindowFunction { partition_by, .. } => Ok(partition_by),
+            Expr::Alias(expr, _) => {
+                // convert &Box<T> to &T
+                match &**expr {
+                    Expr::WindowFunction { partition_by, .. } => Ok(partition_by),
+                    expr => Err(DataFusionError::Execution(format!(
```
Comment (alamb) [inline datafusion/sql/src/utils.rs:487]:
```suggestion
                    expr => Err(DataFusionError::Internal(format!(
```

**Review by alamb — APPROVED** [review]
Looks good to me.

**Pull request closed without merging**